marconi/src/Marconi/Index/MintBurn.hs

```diff
@@ -0,0 +1,310 @@
+        , SQL.toField $ mintAssetAssetName mintAsset
+        , SQL.toField $ mintAssetQuantity mintAsset
+        , SQL.toField $ mintAssetRedeemerIdx mintAsset
+        , SQL.toField $ mintAssetRedeemerData mintAsset
+        ]
+
+type Row = (C.SlotNo, C.Hash C.BlockHeader, C.TxId, C.PolicyId, C.AssetName, C.Quantity, Word64, C.ScriptData)
+
+-- | Input rows must be sorted by C.SlotNo.
+fromRows :: [Row] -> [TxMintEvent]
+fromRows rows =  do
+  rs@(r :| _) <- NE.groupBy ((==) `on` slotNo) rows -- group by SlotNo
+  pure $ TxMintEvent (slotNo r) (hash r) $ do
+    rs' <- NE.groupBy1 ((==) `on` txId) rs -- group by TxId
+    pure (txId r, rowToMintAsset <$> rs')
+  where
+    slotNo = view L._1 :: Row -> C.SlotNo
+    hash = view L._2 :: Row -> C.Hash C.BlockHeader
+    txId = view L._3 :: Row -> C.TxId
+    rowToMintAsset :: Row -> MintAsset
+    rowToMintAsset row = MintAsset (row^.L._4) (row^.L._5) (row^.L._6) (row^.L._7) (row^.L._8)
+
+sqliteSelectByTxIdPolicyId :: SQL.Connection -> (SQL.Query, [NamedParam]) -> C.TxId -> C.PolicyId -> IO [TxMintEvent]
+sqliteSelectByTxIdPolicyId sqlCon (conditions, params) txId policyId =
+  fromRows <$> SQL.queryNamed sqlCon query ([":txId" := txId, ":policyId" := policyId] <> params)
+  where query =
+          " SELECT slotNo, blockHeaderHash, txId, policyId                    \
+          \      , assetName, quantity, redeemerIx, redeemerData              \
+          \   FROM minting_policy_events                                      \
+          \  WHERE txId = :txId AND policyId = :policyId " <> conditions <> " \
+          \  ORDER BY slotNo, txId                                            "
+
+sqliteSelectAll :: SQL.Connection -> (SQL.Query, [NamedParam]) -> IO [TxMintEvent]
+sqliteSelectAll sqlCon (conditions, params) = fromRows <$> SQL.queryNamed sqlCon query params
+  where
+    whereClause = if conditions == "" then "" else " WHERE " <> conditions <> " "
+    query =
+      "   SELECT slotNo, blockHeaderHash, txId, policyId       \
+      \        , assetName, quantity, redeemerIx, redeemerData \
+      \     FROM minting_policy_events                         \
+      \     " <> whereClause <> "                              \
+      \ ORDER BY slotNo, txId                                  "
+
+intervalToWhereClause :: RI.QueryInterval C.ChainPoint -> (SQL.Query, [NamedParam])
+intervalToWhereClause qi = case qi of
+  RI.QEverything -> ("", [])
+  RI.QInterval from to
+    | Just from' <- slotMaybe from, Just to' <- slotMaybe to -> ("slotNo BETWEEN :fromSlot AND :toSlot" , [":fromSlot" := from', ":toSlot" := to'])
+    | Just to' <- slotMaybe to -> ("slotNo <= :toSlot" , [":toSlot" := to'])
+    | otherwise -> ("FALSE", [])
+    where
+      slotMaybe :: C.ChainPoint -> Maybe C.SlotNo
+      slotMaybe = \case
+        C.ChainPoint slotNo _ -> Just slotNo
+        C.ChainPointAtGenesis -> Nothing
+
+groupBySlotNo :: [TxMintEvent] -> [TxMintEvent]
```

Comment:
```suggestion
groupBySlotAndHash :: [TxMintEvent] -> [TxMintEvent]
```
You are grouping over two fields, not just one.

Reply:
8f213ed7a, 9005d2981976a5c5fa3bc60eb87dba848f1c21d7